#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/competitions/desafio-de-prediccion-de-precios-de-portatiles

## Librerias

In [20]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error



## Data

In [2]:
# Load the datasets
train_df = pd.read_csv('data/train.csv')
test_df = pd.read_csv('data/test.csv')
sample_submission_df = pd.read_csv('data/sample_submission.csv')

# Display the first few rows of each dataframe to understand the structure
train_df.head(), test_df.head(), sample_submission_df.head()


(    id  laptop_ID  Company                     Product   TypeName  Inches  \
 0  268       1002     Dell               Inspiron 5567   Notebook    15.6   
 1  347        867     Asus  X541NA (N4200/4GB/1TB/W10)   Notebook    15.6   
 2  353        966  Toshiba           Portege Z30-C-1CW   Notebook    13.3   
 3  578        767     Dell                Alienware 17     Gaming    15.6   
 4   42       1241     Dell              Latitude E7270  Ultrabook    12.5   
 
                   ScreenResolution                                   Cpu  \
 0                         1366x768            Intel Core i7 7500U 2.7GHz   
 1                         1366x768  Intel Pentium Quad Core N4200 1.1GHz   
 2                Full HD 1920x1080            Intel Core i5 6200U 2.3GHz   
 3  IPS Panel 4K Ultra HD 3840x2160           Intel Core i7 7700HQ 2.8GHz   
 4  Full HD / Touchscreen 1920x1080            Intel Core i5 6300U 2.4GHz   
 
     Ram                Memory                      Gpu       OpSy

In [3]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv('data/train.csv')
train

id  laptop_ID  Company                     Product            TypeName  \
0     268       1002     Dell               Inspiron 5567            Notebook   
1     347        867     Asus  X541NA (N4200/4GB/1TB/W10)            Notebook   
2     353        966  Toshiba           Portege Z30-C-1CW            Notebook   
3     578        767     Dell                Alienware 17              Gaming   
4      42       1241     Dell              Latitude E7270           Ultrabook   
..    ...        ...      ...                         ...                 ...   
907    89        973     Acer       Chromebook C738T-C2EJ  2 in 1 Convertible   
908   716       1018  Toshiba           Portege Z30-C-16H            Notebook   
909   517       1071     Acer              Aspire ES1-523            Notebook   
910   891        965     Dell               Latitude 7480           Ultrabook   
911  1179       1313     Dell               Inspiron 3567            Notebook   

     Inches                 ScreenResolution  \
0      15.6                         1366x768   
1      15.6                         1366x768   
2      13.3                Full HD 1920x1080   
3      15.6  IPS Panel 4K Ultra HD 3840x2160   
4      12.5  Full HD / Touchscreen 1920x1080   
..      ...                              ...   
907    11.6   IPS Panel Touchscreen 1366x768   
908    13.3                Full HD 1920x1080   
909    15.6                         1366x768   
910    14.0                Full HD 1920x1080   
911    15.6                         1366x768   

                                      Cpu   Ram                Memory  \
0              Intel Core i7 7500U 2.7GHz   8GB               1TB HDD   
1    Intel Pentium Quad Core N4200 1.1GHz   4GB               1TB HDD   
2              Intel Core i5 6200U 2.3GHz   8GB             256GB SSD   
3             Intel Core i7 7700HQ 2.8GHz  16GB  256GB SSD +  1TB HDD   
4              Intel Core i5 6300U 2.4GHz   8GB             256GB SSD   
..                                    ...   ...                   ...   
907  Intel Celeron Dual Core N3060 1.6GHz   4GB    32GB Flash Storage   
908            Intel Core i5 6200U 2.3GHz   4GB             128GB SSD   
909             AMD A8-Series 7410 2.2GHz   8GB               1TB HDD   
910            Intel Core i7 7600U 2.8GHz   8GB             256GB SSD   
911            Intel Core i7 7500U 2.7GHz   8GB               1TB HDD   

                         Gpu       OpSys  Weight  Price_euros  
0         AMD Radeon R7 M445  Windows 10  2.36kg       749.00  
1      Intel HD Graphics 505  Windows 10     2kg       449.00  
2      Intel HD Graphics 520   Windows 7   1.2kg      1460.00  
3    Nvidia GeForce GTX 1070  Windows 10  4.42kg      2868.99  
4      Intel HD Graphics 520   Windows 7  1.26kg      1713.37  
..                       ...         ...     ...          ...  
907    Intel HD Graphics 400   Chrome OS  1.25kg       389.00  
908    Intel HD Graphics 520  Windows 10   1.2kg      1195.00  
909            AMD Radeon R5  Windows 10   2.4kg       449.00  
910    Intel HD Graphics 620  Windows 10  1.36kg      1775.00  
911       AMD Radeon R5 M430       Linux   2.3kg       805.99  

[912 rows x 14 columns]

In [4]:
test = pd.read_csv('data/test.csv')
test

id  laptop_ID Company                                Product  \
0     181       1098      HP                           Spectre x360   
1     708        330    Acer                               Aspire 5   
2     862       1260    Acer                         Aspire ES1-572   
3    1064       1137      HP                         EliteBook 1040   
4     702       1015      HP                                 ENVY -   
..    ...        ...     ...                                    ...   
386  1281        145  Lenovo                     Legion Y520-15IKBN   
387   524       1195  Lenovo                     IdeaPad Y700-15ISK   
388  1015       1070      HP                                 250 G5   
389  1236        104      HP  15-bw000nv (E2-9000e/4GB/500GB/Radeon   
390  1036        258  Lenovo                         Yoga 920-13IKB   

               TypeName  Inches                           ScreenResolution  \
0             Ultrabook    13.3            IPS Panel 4K Ultra HD 3840x2160   
1              Notebook    15.6                                   1366x768   
2              Notebook    15.6                                   1366x768   
3              Notebook    14.0                          Full HD 1920x1080   
4              Notebook    13.3                IPS Panel Full HD 1920x1080   
..                  ...     ...                                        ...   
386              Gaming    15.6                IPS Panel Full HD 1920x1080   
387              Gaming    15.6                IPS Panel Full HD 1920x1080   
388            Notebook    15.6                                   1366x768   
389            Notebook    15.6                          Full HD 1920x1080   
390  2 in 1 Convertible    13.9  IPS Panel Full HD / Touchscreen 1920x1080   

                              Cpu   Ram     Memory                       Gpu  \
0      Intel Core i7 7500U 2.7GHz  16GB  512GB SSD     Intel HD Graphics 620   
1     AMD A12-Series 9720P 2.7GHz   8GB  256GB SSD         AMD Radeon RX 540   
2      Intel Core i3 6006U 2.0GHz   4GB  500GB HDD     Intel HD Graphics 520   
3      Intel Core i5 6200U 2.3GHz   8GB  256GB SSD     Intel HD Graphics 520   
4      Intel Core i5 7200U 2.5GHz   8GB  256GB SSD     Intel HD Graphics 620   
..                            ...   ...        ...                       ...   
386   Intel Core i7 7700HQ 2.8GHz   8GB  256GB SSD  Nvidia GeForce GTX 1050M   
387   Intel Core i7 6700HQ 2.6GHz  16GB  512GB SSD    Nvidia GeForce GTX 960   
388    Intel Core i5 7200U 2.5GHz   4GB  500GB HDD     Intel HD Graphics 620   
389  AMD E-Series E2-9000e 1.5GHz   4GB  500GB HDD             AMD Radeon R2   
390    Intel Core i7 8550U 1.8GHz   8GB  512GB SSD    Intel UHD Graphics 620   

          OpSys  Weight  
0    Windows 10   1.3kg  
1    Windows 10   2.2kg  
2         Linux   2.4kg  
3     Windows 7  1.43kg  
4    Windows 10  1.34kg  
..          ...     ...  
386       No OS   2.4kg  
387  Windows 10  3.31kg  
388       No OS  1.96kg  
389  Windows 10   2.1kg  
390  Windows 10  1.37kg  

[391 rows x 13 columns]

## Data exploration

In [5]:
train_df.describe(include='all').T

count unique                         top freq         mean  \
id                912.0    NaN                         NaN  NaN   645.091009   
laptop_ID         912.0    NaN                         NaN  NaN   666.192982   
Company             912     19                      Lenovo  208          NaN   
Product             912    475               Inspiron 3567   21          NaN   
TypeName            912      6                    Notebook  522          NaN   
Inches            912.0    NaN                         NaN  NaN    15.011404   
ScreenResolution    912     35           Full HD 1920x1080  346          NaN   
Cpu                 912    104  Intel Core i5 7200U 2.5GHz  142          NaN   
Ram                 912      8                         8GB  434          NaN   
Memory              912     36                   256GB SSD  293          NaN   
Gpu                 912     91       Intel HD Graphics 620  199          NaN   
OpSys               912      9                  Windows 10  746          NaN   
Weight              912    158                       2.2kg   86          NaN   
Price_euros       912.0    NaN                         NaN  NaN  1108.122873   

                         std    min     25%    50%     75%     max  
id                380.313813    0.0  319.75  636.5  980.25  1302.0  
laptop_ID         384.873846    1.0   330.5  673.5   998.5  1320.0  
Company                  NaN    NaN     NaN    NaN     NaN     NaN  
Product                  NaN    NaN     NaN    NaN     NaN     NaN  
TypeName                 NaN    NaN     NaN    NaN     NaN     NaN  
Inches              1.411744   10.1    14.0   15.6    15.6    18.4  
ScreenResolution         NaN    NaN     NaN    NaN     NaN     NaN  
Cpu                      NaN    NaN     NaN    NaN     NaN     NaN  
Ram                      NaN    NaN     NaN    NaN     NaN     NaN  
Memory                   NaN    NaN     NaN    NaN     NaN     NaN  
Gpu                      NaN    NaN     NaN    NaN     NaN     NaN  
OpSys                    NaN    NaN     NaN    NaN     NaN     NaN  
Weight                   NaN    NaN     NaN    NaN     NaN     NaN  
Price_euros       714.597741  174.0   589.0  949.0  1458.5  6099.0

In [6]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                912 non-null    int64  
 1   laptop_ID         912 non-null    int64  
 2   Company           912 non-null    object 
 3   Product           912 non-null    object 
 4   TypeName          912 non-null    object 
 5   Inches            912 non-null    float64
 6   ScreenResolution  912 non-null    object 
 7   Cpu               912 non-null    object 
 8   Ram               912 non-null    object 
 9   Memory            912 non-null    object 
 10  Gpu               912 non-null    object 
 11  OpSys             912 non-null    object 
 12  Weight            912 non-null    object 
 13  Price_euros       912 non-null    float64
dtypes: float64(2), int64(2), object(10)
memory usage: 99.9+ KB


## Data processing

In [7]:
# Convert 'Ram' to numeric by removing 'GB' and converting to int
train_df['Ram'] = train_df['Ram'].str.replace('GB', '').astype(int)
test_df['Ram'] = test_df['Ram'].str.replace('GB', '').astype(int)

In [8]:
# Convert 'Weight' to numeric by removing 'kg' and converting to float
train_df['Weight'] = train_df['Weight'].str.replace('kg', '').astype(float)
test_df['Weight'] = test_df['Weight'].str.replace('kg', '').astype(float)

In [9]:
# Split 'ScreenResolution' into width and height
train_df[['Resolution_Width', 'Resolution_Height']] = train_df['ScreenResolution'].str.extract(r'(\d+)x(\d+)').astype(int)
test_df[['Resolution_Width', 'Resolution_Height']] = test_df['ScreenResolution'].str.extract(r'(\d+)x(\d+)').astype(int)

In [10]:
# One-Hot Encoding for categorical variables
train_df = pd.get_dummies(train_df, columns=['Company', 'TypeName', 'OpSys'])
test_df = pd.get_dummies(test_df, columns=['Company', 'TypeName', 'OpSys'])

In [11]:
# Ensure both train and test have the same dummy columns
train_df, test_df = train_df.align(test_df, join='left', axis=1, fill_value=0)

In [12]:
# # Drop unnecessary columns
# train_df = train_df.drop(columns=['laptop_ID', 'ScreenResolution'])
# test_df = test_df.drop(columns=['laptop_ID', 'ScreenResolution'])

In [13]:
train_df.head()

id  laptop_ID                     Product  Inches  \
0  268       1002               Inspiron 5567    15.6   
1  347        867  X541NA (N4200/4GB/1TB/W10)    15.6   
2  353        966           Portege Z30-C-1CW    13.3   
3  578        767                Alienware 17    15.6   
4   42       1241              Latitude E7270    12.5   

                  ScreenResolution                                   Cpu  Ram  \
0                         1366x768            Intel Core i7 7500U 2.7GHz    8   
1                         1366x768  Intel Pentium Quad Core N4200 1.1GHz    4   
2                Full HD 1920x1080            Intel Core i5 6200U 2.3GHz    8   
3  IPS Panel 4K Ultra HD 3840x2160           Intel Core i7 7700HQ 2.8GHz   16   
4  Full HD / Touchscreen 1920x1080            Intel Core i5 6300U 2.4GHz    8   

                 Memory                      Gpu  Weight  ...  \
0               1TB HDD       AMD Radeon R7 M445    2.36  ...   
1               1TB HDD    Intel HD Graphics 505    2.00  ...   
2             256GB SSD    Intel HD Graphics 520    1.20  ...   
3  256GB SSD +  1TB HDD  Nvidia GeForce GTX 1070    4.42  ...   
4             256GB SSD    Intel HD Graphics 520    1.26  ...   

   TypeName_Workstation  OpSys_Android  OpSys_Chrome OS  OpSys_Linux  \
0                 False          False            False        False   
1                 False          False            False        False   
2                 False          False            False        False   
3                 False          False            False        False   
4                 False          False            False        False   

   OpSys_Mac OS X  OpSys_No OS  OpSys_Windows 10  OpSys_Windows 10 S  \
0           False        False              True               False   
1           False        False              True               False   
2           False        False             False               False   
3           False        False              True               False   
4           False        False             False               False   

   OpSys_Windows 7  OpSys_macOS  
0            False        False  
1            False        False  
2             True        False  
3            False        False  
4             True        False  

[5 rows x 47 columns]

In [14]:
test_df.head()

id  laptop_ID         Product  Inches                 ScreenResolution  \
0   181       1098    Spectre x360    13.3  IPS Panel 4K Ultra HD 3840x2160   
1   708        330        Aspire 5    15.6                         1366x768   
2   862       1260  Aspire ES1-572    15.6                         1366x768   
3  1064       1137  EliteBook 1040    14.0                Full HD 1920x1080   
4   702       1015          ENVY -    13.3      IPS Panel Full HD 1920x1080   

                           Cpu  Ram     Memory                    Gpu  Weight  \
0   Intel Core i7 7500U 2.7GHz   16  512GB SSD  Intel HD Graphics 620    1.30   
1  AMD A12-Series 9720P 2.7GHz    8  256GB SSD      AMD Radeon RX 540    2.20   
2   Intel Core i3 6006U 2.0GHz    4  500GB HDD  Intel HD Graphics 520    2.40   
3   Intel Core i5 6200U 2.3GHz    8  256GB SSD  Intel HD Graphics 520    1.43   
4   Intel Core i5 7200U 2.5GHz    8  256GB SSD  Intel HD Graphics 620    1.34   

   ...  TypeName_Workstation  OpSys_Android  OpSys_Chrome OS  OpSys_Linux  \
0  ...                 False              0            False        False   
1  ...                 False              0            False        False   
2  ...                 False              0            False         True   
3  ...                 False              0            False        False   
4  ...                 False              0            False        False   

   OpSys_Mac OS X  OpSys_No OS  OpSys_Windows 10  OpSys_Windows 10 S  \
0           False        False              True               False   
1           False        False              True               False   
2           False        False             False               False   
3           False        False             False               False   
4           False        False              True               False   

   OpSys_Windows 7  OpSys_macOS  
0            False        False  
1            False        False  
2            False        False  
3             True        False  
4            False        False  

[5 rows x 47 columns]

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [18]:
X = train_df.drop(columns=['Price_euros'])
y = train_df['Price_euros']

### 2. Dividir X_train, X_test, y_train, y_test

In [21]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

### 3. Crear y entrenar el modelo


In [22]:
# Entrenar el modelo de regresión lineal
model = LinearRegression()
model.fit(X_train, y_train)

ValueError: could not convert string to float: 'X540UA-DM186 (i3-6006U/4GB/1TB/FHD/Linux)'

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de MAE.

In [17]:
# Evaluar el modelo utilizando MAE
mae = mean_absolute_error(y_val, y_pred)

mae

NameError: name 'mean_absolute_error' is not defined

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

In [23]:
# Realizar predicciones en el conjunto de prueba
test_pred = model.predict(test_df.drop(columns=['Price_euros']))

# Crear el dataframe de resultados
submission_df = test_df[['id']].copy()
submission_df['Price_euros'] = test_pred

# Asegurarnos de que el formato sea igual al de 'sample_submission.csv'
submission_df = submission_df[['id', 'Price_euros']]

# Guardar el archivo de envío
submission_file_path = '/mnt/data/submission.csv'
submission_df.to_csv(submission_file_path, index=False)

submission_file_path


ValueError: could not convert string to float: 'Spectre x360'

In [ ]:
# Crear una columna 'id' basada en los índices del conjunto de prueba
test_df['id'] = test_df.index

# Realizar predicciones en el conjunto de prueba
test_pred = model.predict(test_df.drop(columns=['id', 'Price_euros']))

# Crear el dataframe de resultados
submission_df = test_df[['id']].copy()
submission_df['Price_euros'] = test_pred

# Asegurarnos de que el formato sea igual al de 'sample_submission.csv'
submission_df = submission_df[['id', 'Price_euros']]

# Guardar el archivo de envío
submission_file_path = '/mnt/data/submission.csv'
submission_df.to_csv(submission_file_path, index=False)

submission_file_path


### 2. Carga los datos de `test.csv` para predecir.

**¿Por qué puede dar error?** 

IMPORTANTE: APLICAR A ESTOS DATOS LO MISMO QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [15]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [16]:
predictions = model.predict(X_pred)
predictions

NameError: name 'model' is not defined

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [ ]:
sample = pd.read_csv("Dataset/sample_submission.csv") # Esta es mi ruta del archivo, la vuestra puede ser otra

In [ ]:
sample.head()

In [ ]:
sample.shape

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [ ]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("")


In [ ]:
chequeator(submission) # submission es el nombre que le habríamos puesto a nuestro .csv con los valores que me salieron en la predicción

You're ready to submit!
